### PR TITLE
MIDI for light: Implement new Active deck heuristic

### LINKED ISSUE #11322: Midi for light gets stuck not detecting deck change
### Bug Description

Sometimes it is possible that the [`Midi_for_light` MIDI script](https://github.com/mixxxdj/mixxx/wiki/midi_clock_output) gets stuck, and stops detecting deck changes, thereby stopping beats from transmitting.

A reliable way to reproduce it:
1. Load 2 tracks into deck A/B and set the crossfader to deck B
2. Enable the MIDI script in the settings now
3. Start playing the track on deck B, note that beat events are transmitted
4. Press Cue on deck B to stop the track
5. Set the crossfader to deck A, note that this does not transmit a deck change event
6. Start playing the track on deck A, note that no beat events are transmitted

### Version

2.3.3

### OS

Windows 11

### LINKED ISSUE #13138: MIDI for light: Active deck heuristic could be improved
### Feature Description

The MIDI for light mapping currently only seems to use the crossfader to determine the active deck, which as a heuristic already breaks down if e.g. the crossfader is set to "through" and the deck faders are used to mix. The active deck shown in Mixxx's title bar handles this correctly, perhaps there is a way to access this info from the mapping script? If not, we should at least be able to improve the heuristic.

### PR DESCRIPTION
**Pull request by ekordas**
Detects the active Deck by sorting the Decks by their volumes.
fixes #11322, closes #13138

### CODE REVIEW

File: res/controllers/Midi_for_light-scripts.js [hunk 1]
```diff
@@ -41,6 +41,9 @@ var enable_vu_right_average_max = false; // set to false if you not need VU righ
 var enable_vu_right_average_fit = true; // set to false if you not need VU right average fit
 var enable_vu_right_current_meter = false; // set to false if you not need VU right current meter
 var enable_vu_right_average_meter = false; // set to false if you not need VU right average meter
+var deck_ending_time = 15; // set a time (in soconds) in wich the playing track is considderd ending
```
Comment (fwcd) [inline res/controllers/Midi_for_light-scripts.js:44]:
```suggestion
var deck_ending_time = 15; // set a time (in seconds) in which the playing track is considered to be ending
```

File: res/controllers/Midi_for_light-scripts.js [hunk 2]
```diff
@@ -65,6 +68,7 @@ if (enable_vu_mono_current === true || enable_vu_mono_average_min === true || en
 } else {
     var enable_vu_meter_global = false; // set to false if you not need complete VU-Meter
 }
+var last_mtp_playpositon = -1
```
Comment (fwcd) [inline res/controllers/Midi_for_light-scripts.js:71]:
```suggestion
var last_mtp_playposition = -1
```

File: res/controllers/Midi_for_light-scripts.js [hunk 3]
```diff
@@ -504,6 +462,13 @@ midi_for_light.sendMidiMtcFullFrame = function(value, group, control) { // sends
     var duration = engine.getValue(group, "track_samples") / engine.getValue(group, "track_samplerate") / 2;
     var PlayPositionRest = duration * engine.getValue(group, "playposition");
 
+    //Prevent outputputting the same Position twice
+    var current_mtp_playpositon = Math.floor(PlayPositionRest*25);
+    if (current_mtp_playpositon == last_mtp_playpositon) {
+        return;
+    }
+    last_mtp_playpositon = current_mtp_playpositon;
```
Comment (fwcd) [inline res/controllers/Midi_for_light-scripts.js:470]:
```suggestion
    // Prevent outputting the same Position twice
    var current_mtp_playposition = Math.floor(PlayPositionRest*25);
    if (current_mtp_playposition == last_mtp_playposition) {
        return;
    }
    last_mtp_playposition = current_mtp_playposition;
```

File: res/controllers/Midi_for_light-scripts.js [hunk 4]
```diff
@@ -401,99 +447,11 @@ midi_for_light.vuMeter = function() { // read, calculate and send vu-meter value
...
-    midi.sendShortMsg(0x7F + midi_channel, 0x30, 0x0); // note C off with value 0
-};
-
-midi_for_light.volumeBeatOnDelay = function() { // allow deck change with volume after 3 second fader do nothing
-    if (midi_for_light.volumebeat_on_delay_timer) {
-        engine.stopTimer(midi_for_light.volumebeat_on_delay_timer);
-        midi_for_light.volumebeat_on_delay_timer = undefined;
-    }
-    midi_for_light.volumebeat = true;
+    midi_for_light.calculateDeckPriority()
```
Comment (fwcd) [inline res/controllers/Midi_for_light-scripts.js:450]:
```suggestion
    midi_for_light.calculateDeckPriority();
```

File: res/controllers/Midi_for_light-scripts.js [hunk 5]
```diff
@@ -107,14 +111,61 @@ midi_for_light.shutdown = function(id) { // called when the MIDI device is close
...
+            midi_for_light.decks[i].priority = 0.0;
+            continue;
+        }
+
+        midi_for_light.decks[i].priority = engine.getParameter(channel, "volume") * crossfader_factors[engine.getValue(channel, "orientation")];
+
+        // Decrease Priority of ending Tracks
+        var duration = engine.getValue(channel, "duration");
+        var PlayPosition = duration * engine.getValue(channel, "playposition");
+        if (duration - PlayPosition < deck_ending_time) {
```
Comment (fwcd) [inline res/controllers/Midi_for_light-scripts.js:142]:
```suggestion
        var playPosition = duration * engine.getValue(channel, "playposition");
        if (duration - playPosition < deck_ending_time) {
```

for consistency. I would, by the way, also recommend using `const` or `let` in new code where possible, even if other parts of the script still use `var`.

**Review by fwcd — COMMENTED** [review]
Thanks! Very happy to see this being worked on, improvements to the script are very much appreciated and this was an issue I've seen in the past too (even when using the crossfader the detection always seemed somewhat unreliable).

A couple of style nits below, hopefully I'll find some time to do some actual testing of this script.

**Comment by fwcd** [comment]
Ah, I think you may want to install the pre-commit hook too, that should take care of most of the formatting issues.

**Comment by JoergAtGithub** [comment]
Welcome at Mixxx!
As a first-time contributor we need you to sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and comment here when you have done so. It gives us permission to distribute your contribution under the GPL v2 or later license and the Apple Mac App Store. It is also helpful for us to have contact information for contributors in case we may need it in the future.

**Comment by ekordas** [comment]
@JoergAtGithub Done

**Comment by JoergAtGithub** [comment]
The pre-commit check is failing. The best way to fix pre-commit issues is to install pre-commit locally on your system, as described here: https://github.com/mixxxdj/mixxx/wiki/Using-Git#set-up-automatic-code-checking
Than it will fix these issues automatically before every commit.
Alternatively, you can download the pre-commit.patch file from the artifacts of this PR
grafik
unzip it, and apply it using:
`git apply pre-commit.patch`

**Comment by ekordas** [comment]
I used that pre-commit thing, but it seems it did not find everything on my end... is there something I need to configure?

**Comment by ekordas** [comment]
Sorry for flooding commits here. Now I used that downloaded patch. I hope this will fix it now.

**Comment by fwcd** [comment]
Would you mind fixing the conflicts? I think they're coming from a bunch of typo fixes in https://github.com/mixxxdj/mixxx/pull/13567

**Review by fwcd — APPROVED** [review]
Just got a chance to try it out, this is great and already works much better than the current logic! I usually only use the channel faders and keep the crossfader disabled, so having the logic account for this makes it a lot more useful.

The code looks good, but since it is more than a trivial fix, I would be hesitant to merge it to 2.4. Could you retarget it to 2.5?

**Comment by ekordas** [comment]
Is this how it is supposed to work? I never changed the Pull request target ^^

**Comment by fwcd** [comment]
It seems to merge cleanly, but your branch is still based on 2.4, so it would be great if you could rebase your branch to 2.5, so we get another CI build:

```sh
# Make sure 2.4 and 2.5 are pulled and up-to-date
git checkout <your branch>
git rebase 2.4 --onto 2.5
git push --force-with-lease
```

**Comment by ekordas** [comment]
OK, I will do it once i am at home

**Comment by ekordas** [comment]
I have still no Idea why pre-commit still does not fail on my end in some cases. But anyways it seems some things went wrong rebasing the branch. Im now building 2.5 to test if all still works...

**Comment by ekordas** [comment]
tested in 2.5, should work now.

**Comment by fwcd** [comment]
Tested and works. Cool, thanks!

**Comment by ekordas** [comment]
Thanks for merging. The issues mentioned in the description can be closed.

**Comment by fwcd** [comment]
Done!

**Pull request merged**